packages/next/client/index.js

```diff
@@ -323,14 +329,70 @@ function renderReactElement(reactEl, domEl) {
 
   if (onPerfEntry && ST) {
     try {
-      const observer = new PerformanceObserver(list => {
-        list.getEntries().forEach(onPerfEntry)
-      })
-      // Start observing paint entry types.
-      observer.observe({
-        type: 'paint',
-        buffered: true,
-      })
+      function getSupportedTypes(types) {
+        if (PerformanceObserver && PerformanceObserver.supportedEntryTypes) {
+          return types.filter(type =>
+            PerformanceObserver.supportedEntryTypes.includes(type)
+          )
+        } else {
+          return []
+        }
+      }
+
+      function observeSupportedTypes(entryTypes) {
+        const supportedTypes = getSupportedTypes(entryTypes)
+
+        /**
+         * Layout shift has special handling as this will be sent
+         * cumulative layout shift.
+         */
+        if (supportedTypes.includes('layout-shift')) {
+          // Stores the current layout shift score for the page.
+          let cumulativeLayoutShiftScore = 0
+          const observer = new PerformanceObserver(list => {
+            for (const entry of list.getEntries()) {
+              // Only count layout shifts without recent user input.
+              if (!entry.hadRecentInput) {
+                cumulativeLayoutShiftScore += entry.value
+              }
+            }
+          })
+          observer.observe({
+            type: 'layout-shift',
+            buffered: true,
+          })
+          document.addEventListener('visibilitychange', () => {
+            if (document.visibilityState === 'hidden') {
+              // Force any pending records to be dispatched.
+              observer.takeRecords()
+              observer.disconnect()
+              onPerfEntry({
+                name: 'cumulative-layout-shift',
+                value: cumulativeLayoutShiftScore,
+              })
+            }
+          })
+        }
+
+        // For all other entry types, we send values to relayer as given by PerformanceObserver
+        supportedTypes.forEach(type => {
+          if (type === 'layout-shift') {
+            return
+          }
+          const observer = new PerformanceObserver(list => {
+            list.getEntries().forEach(onPerfEntry)
+          })
+          observer.observe({
+            type,
+            buffered: true,
+          })
+        })
+      }
+      observeSupportedTypes([
+        'paint',
+        'largest-contentful-paint',
```

Comment:
LCP is similar to CLS in that you can't necessarily get its value from a single entry. Multiple `largest-contentful-paint` entries can be dispatched, and only the last one dispatched is important: https://web.dev/lcp/#measure-lcp-in-javascript

Note: I can't quite tell from this code, but I'm assuming you're not accounting for that (or at least I can't see from the code in this PR).

Reply:
done